File: w3/E-Fibonacci Sequence.cpp
// URAL 1133
// 用C++去解方程，应该是误差的问题，wa在第12个案例上了，最后改用快速幂去二分就好了，看py的代码
#include <iostream>
#include <vector>
using namespace std;

typedef long long LL;
typedef vector<vector<LL> > Matrix;
Matrix Unit(2);

Matrix mul(const Matrix& a, const Matrix& b) {
	Matrix ans(a[0].size(), vector<LL>(b.size()));

	for (int i = 0; i < a[0].size(); ++i) {
		for (int j = 0; j < b.size(); ++j) {
			LL sum = 0;
			for (int k = 0; k < b.size(); ++k) {
				sum += a[k][i] * b[j][k];
			}
			ans[i][j] = sum;
		}
	}

	return ans;
}

Matrix cal(int n) {
	Matrix ans(Unit);
	for (int i = 1; i < n; ++i) 
		ans = mul(ans, Unit);
	return ans;
}

int main() {
	LL i, j, n, Fi, Fj;
	cin >> i >> Fi >> j >> Fj >> n;
	Unit[0].push_back(1);	Unit[0].push_back(1);
	Unit[1].push_back(1);	Unit[1].push_back(0);

	LL X = min(min(i, j), n);
	i -= (X - 1 b 	j -= (X - 1);
	n -= (X - 1);
	
	Matrix I = cal(i);
	Matrix J = cal(j);
	LL a1 = I[0][0], b1 = I[0][1], a2 = J[0][0], b2 = J[0][1];
	LL F1 = (Fi*b2 - Fj*b1) / (a1*b2 - a2*b1);
	LL F0 = (Fi*a2 - Fj*a1) / (a2*b1 - a1*b2);

//cout << a1 << ' ' << b1 << ' ' << a2 << ' ' << b2 << ' ' << F1 << ' ' << F0 << endl;
	Matrix N = cal(n);
	cout << N[0][0]*F1 + N[0][1]*F0 << endl;

	return 0;
}

File: w3/E-Fibonacci Sequence.py
# URAL 1133
Unit = [[1, 1], [1, 0]]

def mul(a, b):
	ans = [[0 for i in range(0, len(a))] for j in range(0, len(b))]
	for i in range(0, len(a)):
		for j in range(0, len(b)):
			for k in range(0, len(b)):
				ans[i][j] += a[i][k] * b[k][j]
	return ans

def POW(n):
	a = Unit
	b = [[1, 0], [0, 1]]
	while n > 0:
		if n & 1:
			b = mul(a, b)
		a = mul(a, a)
		n >>= 1
	return b

def cal(i, Fi, j, Fj, n):
	mat = POW(j-1)
	[Left, Right] = [int(-2*1e9), int(2*1e9)]
	while Left <= Right:
		mid = (Left + Right) >> 1
		x = mat[0][0] * mid + mat[0][1] * Fi
		if x > Fj:
			Right = mid - 1
		elif x < Fj:
			Left = mid + 1
		else:
			break
	
	if n > 0:
		mat = POW(n - 1)
		print(mid*mat[0][0] + Fi*mat[0][1])
	else:
		[Fa, Fb] = [mid, Fi]
		for i in range(n, 0):
			[Fa, Fb] = [Fb, Fa - Fb]
		print(Fb)


if __name__ == '__main__':
	[i, Fi, j, Fj, n] = [int(i) for i in input().split(' ')]
	if i > j:
		[i, j] = [j, i]
		[Fi, Fj] = [Fj, Fi]
	j -= i
	n -= i
	i = 0
	cal(i, Fi, j, Fj, n)
